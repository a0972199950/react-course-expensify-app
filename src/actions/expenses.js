import database from "../firebase/firebase";

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense    
});

export const startAddExpense = (expenseData = {}) => {
    // 藉由在store裡面傳入redux-thunk，我們可以讓action不要return一個object，而是return一個函數進去store
    // redux-thunk會把dispatch傳進我們return的函數作為第一個參數來讓我們使用
    return (dispatch) => {
        // 解構expenseData
        const {description = "", note = "", amount = 0, createdAt = 0} = expenseData;
        const expense = { description, note, amount, createdAt };
        // push資料進firebase 
        // 這邊的return會return一個Promise，是為了給tests用的
        // 之後會在tests裡面再對上面的Promise連接.then()
        return database
            .ref("expenses")
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }))
            })
    }
}


export const removeExpense = (id = undefined) => ({
    type: "REMOVE_EXPENSE",
    id
});


export const startRemoveExpense = (id = undefined) => {
    return (dispatch) => {
        return database
            .ref("expenses/" + id)
            .remove()
            .then(() => {
                dispatch(removeExpense(id));
            })
            .catch((e) => {
                console.log("remove failed", e);
            });

    };
};


export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});


export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database
            .ref("expenses/" + id)
            .update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            })
            .catch((e) => {
                console.log("update failed", e);
            });
    };
};


// 將從database拿到的資料寫到redux裡面
export const setExpenses = (expenses) => {
    return {
        type: "SET_EXPENSES",
        expenses
    }
};


// 從database獲取既存的資料
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref("expenses").once("value").then((snapshot) => {
            const expensesArray = [];

            snapshot.forEach((childSnapshot) => {
                expensesArray.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expensesArray));

        }).catch((e) => {
            console.log("fatch data from database failed", e);
        });
    };
};