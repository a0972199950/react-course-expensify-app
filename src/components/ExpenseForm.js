import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);

        this.state = {
            description: props.expenseToEdit ? props.expenseToEdit.description : "",
            note: props.expenseToEdit ? props.expenseToEdit.note : "",
            amount: props.expenseToEdit ? props.expenseToEdit.amount : "",
            createdAt: props.expenseToEdit ? moment(props.expenseToEdit.createdAt) : moment(),
            calendarFocused: false,
            error: ""
        };
    }

    onDescriptionChange = (event) => {
        const description = event.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (event) => {
        const amount = event.target.value;
        // 這表達式到底是三小也太複雜了......
        if(amount.match(/^\d*(\.{0,0})?$/)){
            this.setState(() => ({ amount }));
        }
    }

    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (event) => {
        event.preventDefault();

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({ error: "Please provide the description and amount" }));
        } else{
            this.setState(() => ({ error: "" }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseInt(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        value={this.state.description}
                        placeholder="description"
                        onChange={this.onDescriptionChange}
                        // autoFocus的作用是讓網頁剛進入時就FOCUS在某個input上面，
                        // 這樣user就不用再把游標移過去
                        autoFocus
                    />

                    <input 
                        type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="note(optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>{this.props.expenseToEdit ? "Edit Expense" : "Add Expense"}</button>
                </form>
            </div>
        );        
    }
}