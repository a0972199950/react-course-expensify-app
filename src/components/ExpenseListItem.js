import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({description, amount, createdAt, id}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
        TWD {numeral(amount).format("$0,0")}
        -
        {moment(createdAt).format("YYYY/MM/DD HH:mm:ss")}</p>
    </div>
);



export default ExpenseListItem;