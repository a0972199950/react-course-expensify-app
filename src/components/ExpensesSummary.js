import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import expensesTotal from "../selectors/expenses-total";


export const ExpensesSummary = (props) => {
    return (
        <div>
            <p>
                Viewing {props.count} expenses totalling {numeral(props.total).format("$0,0")}
            </p>
        </div>
    )
};

const mapStateToProps = (state) => {
    const selectedExpenses = selectExpenses(state.expenses, state.filters);
    return{
        count: selectedExpenses.length,
        total: expensesTotal(selectedExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);