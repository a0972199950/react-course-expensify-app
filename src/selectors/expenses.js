import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {

        const startDateMatch = !startDate || moment(expense.createdAt).isSameOrAfter(startDate, "date");
        const endDateMatch = !endDate || moment(expense.createdAt).isSameOrBefore(endDate, "date");
        const textMatch = !text || expense.description.includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((last, next) => {
        switch(sortBy){
            case "amount":
                return (last.amount < next.amount) ? -1 : 1;

            case "date": 
                return (last.createdAt < next.createdAt) ? -1 : 1;

            default:
                return 0;
        };
    });

};