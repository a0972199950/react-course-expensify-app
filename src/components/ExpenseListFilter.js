import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";




export class ExpenseListFilter extends React.Component{
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value);            
    }

    onSortByChange = (event) => {
        const sortBy = event.target.value;
        if(sortBy === "date") this.props.sortByDate()
        else if(sortBy === "amount") this.props.sortByAmount();
    }

    render(){
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        {/* value會不停地從state讀取資料並顯示，所以當state的資料改變時，它也會同步改變
                        但defaultValue不會持續讀取，只會在一開始時讀取一次。
                        所以若想持續追蹤資料，則必須使用value；同時加上onChange來讓輸入框可以被編輯 */}
                        <input 
                            type="text" 
                            className="text-input"
                            placeholder="Search expense"
                            value={this.props.filters.text} 
                            onChange={this.onTextChange} 
                        />
                    </div>

                    <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filters.sortBy} 
                            onChange={this.onSortByChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>

                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate} 
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange} 
                            isOutsideRange= {() => false}
                            numberOfMonths={1}
                            showClearDates={true}
                        />
                    </div>
                </div>
            
                
        
                
        
                
            </div>
        );
    }
    
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => {dispatch(setTextFilter(text))},
    sortByDate: () => {dispatch(sortByDate())},
    sortByAmount: () => {dispatch(sortByAmount())},
    setStartDate: (startDate) => {dispatch(setStartDate(startDate))},
    setEndDate: (endDate) => {dispatch(setEndDate(endDate))}
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);

