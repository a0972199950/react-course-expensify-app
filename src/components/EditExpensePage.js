import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component{
    onSubmit = (expenseToEdit) => {
        this.props.startEditExpense(this.props.expenseToEdit.id, expenseToEdit)
        this.props.history.push("/");
    }

    handleDeleteExpense = () => {
        this.props.startRemoveExpense(this.props.expenseToEdit.id);
        this.props.history.push("/");
    }

    render(){
        return (
            <div>
                <h3>Edit Expense</h3>
                <ExpenseForm 
                    expenseToEdit={this.props.expenseToEdit}
                    onSubmit={this.onSubmit} 
                />
                <button 
                    onClick={this.handleDeleteExpense}
                >
                    Delete
                </button>
            </div>
        );
    }
        
};




// connect函數除了會把props傳給要連結的組件外，也會把尚未包含state的props傳給mapStateToProps
const mapStateToProps = (state, props) => {
    return {
        expenseToEdit: state.expenses.find((expense) => expense.id === props.match.params.id)        
    }
};

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, updates) => {
        dispatch(startEditExpense(id, updates))
    },
    startRemoveExpense: (id) => {
        dispatch(startRemoveExpense(id))
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);