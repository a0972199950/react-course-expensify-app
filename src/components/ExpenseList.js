import React from "react";
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
    <div>
        {props.expenses.length > 0 ? (
            props.expenses.map((expense) => (
                <ExpenseListItem key={expense.id} {...expense} />
            ))
        ) : (
            <p>No expenses exist.</p>
        )}
        {}
    </div>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// connect()預設吃的參數是一個函數。
// connect()會自動將Provider的store的state傳入該函數，並且該函數必須要回傳一個物件，內容是你想要存取的tore內容(所以不會全部都能存取)
// 而呼叫connect()的時候丟給它別的東西，那就是你想要connect的組件
// 很奇怪的api吧 0.0
export default connect(mapStateToProps)(ExpenseList);

