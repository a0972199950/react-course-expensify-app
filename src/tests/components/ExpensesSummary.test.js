import { ExpensesSummary } from "../../components/ExpensesSummary";
import { shallow } from "enzyme";
import React from "react";
import expensesTotal from "../../selectors/expenses-total";
import expenses from "../fixtures/expense";

let count, total, wrapper;
beforeEach(() => {
    wrapper = shallow(<ExpensesSummary />);
})

test("should render count and total with single expense", () => {
    const expense = [expenses[0]];
    count = expense.length;
    total = expensesTotal(expense);

    wrapper.setProps({ count, total });
    expect(wrapper).toMatchSnapshot();
});


test("should render count and total with multiple expenses", () => {
    count = expenses.length;
    total = expensesTotal(expenses);

    wrapper.setProps({ count, total });
    expect(wrapper).toMatchSnapshot();
});
