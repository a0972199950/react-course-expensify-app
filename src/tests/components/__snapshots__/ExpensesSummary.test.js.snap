// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render count and total with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling 
    $0
  </h1>
</div>
`;

exports[`should render count and total with single expense 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling 
    $0
  </h1>
</div>
`;
